libres/lib/job_queue/job_queue.cpp

```diff
@@ -538,81 +554,69 @@ static bool job_queue_check_node_status_files(const job_queue_type *job_queue,
     return false;
 }
 
-static void *job_queue_run_DONE_callback(void *arg) {
-    arg_pack_type *arg_pack = arg_pack_safe_cast(arg);
-    job_queue_type *job_queue =
-        (job_queue_type *)arg_pack_iget_ptr(arg_pack, 0);
-    int queue_index = arg_pack_iget_int(arg_pack, 1);
-    job_list_get_rdlock(job_queue->job_list);
-    {
-        job_queue_node_type *node =
-            job_list_iget_job(job_queue->job_list, queue_index);
-        bool OK = job_queue_check_node_status_files(job_queue, node);
+static void job_queue_run_DONE_callback(job_queue_type *job_queue,
+                                        job_queue_node_type *node) {
+    JobListReadLock read_lock(
+        job_queue->job_list); // Keep in mind that this runs on another thread
+                              // than the code triggering it, so we need this
 
-        if (OK)
-            OK = job_queue_node_run_DONE_callback(node);
+    // There is a small timeslot in which status may change after we decide to
+    // run this handler, and before we get the readlock above. Handle it...
+    if (JOB_QUEUE_DONE != job_queue_node_get_status(node)) {
+        logger->warning("Job {}: expected status {} got {}",
+                        job_queue_node_get_name(node), JOB_QUEUE_DONE,
+                        job_queue_node_get_status(node));
+        return;
+    }
```

Comment:
If `run_handlers` locks for reading, we can't mutate the status here… So the hope is the `job_queue_loop` will re-run this handler?

So if this is fine, we should log `info`, if it's not we should crash.

Reply:
Note that there is a second TOCTOU here 😂 which means you could get a log message indicating _nothing_ went wrong.

Reply:
I probably don't grasp 100% what you mean, but AFAICS this module has never held any writelock when changing status of a node, of which the robustness is another discussion. This method is the outmost scope under our control in the thread, and grabbing the readlock (using `JobListReadLock`) prevents anyone from mutating the list (FWIW) like the original code did.

The difference is that this code "asserts" that status is as expected. As you pointed out in a deleted comment, there is in fact a possible race-condition between the check and the printout which becomes possible because we don't protect status-changes with a lock, as commented above. Should be avoided now.

Returning without mutating state and leave this to other code seems reasonable here. Logging with info-level is fine.